### PR TITLE
Updating Firefly-dependent areas to match API changes and PyPI packaging

### PR DESCRIPTION
**Pull request by mtryan83**
## PR Summary

This PR aims to reconcile the changes to Firefly introduced since version 4.0.0 of yt was released. In summary, Firefly tweaked the `Reader` and `ParticleGroup` APIs and the installation of Firefly from PyPI installs as `firefly` (instead of `Firefly` or `Firefly-vis`, the devs were able to get control of the package name).  I've updated the documentation, including that (I believe) Firefly also now prefers their binary `.ffly` format.

I'm very new to Firefly and yt, so I would greatly appreciate it if e.g. @agurvich would double check.

I believe this counts as a fix/update for #3415 

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by mtryan83** [comment]
pre-commit.ci autofix

File: setup.cfg
Comment (neutrinoceros) [inline setup.cfg:84]: I would like to keep this as bullet-proof as possible, so that yt+firefly keeps working in CI even when API changes upstream. AFAIC I'm very happy with the current workflow where this kind of PR helps us migrate to newer API, but we do not depend.
```suggestion
    firefly>=3.2.0,<4.0.0
```

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:706]: this particular change is not 100% backward compatible for yt users, we would also need to capture `JSONdir` being passed as a keyword argument and raise a deprecation warning when it's caugh (use `yt._maintainance.deprecation.issue_deprecation_warning`).
This also implies adding a default value to `datadir` (`None`)... well it gets messy pretty quickly. Let me know if you need more detailed instructions. I can also jump in and do this bit myself if you'd like.

File: yt/utilities/on_demand_imports.py
Comment (neutrinoceros) [inline yt/utilities/on_demand_imports.py:455]: better to keep the name of the class in sync too
```suggestion
class firefly_imports(OnDemand):
```

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
Hi @mtryan83, thanks a lot for doing this !
Indeed, it'd be great if @agurvich could sign off here.
In the mean time, here are a couple comments, one of which is crucial (API change)

File: yt/utilities/on_demand_imports.py
Comment (neutrinoceros) [reply yt/utilities/on_demand_imports.py:455]: (note that you'll need to update line 469 as well)

File: yt/data_objects/data_containers.py
Comment (mtryan83) [reply yt/data_objects/data_containers.py]: I think 12c54cc covers this? I followed the approach from yt/data_objects/image_array.py for the notation. I put in 4.2 as the removal version, but that was just a guess.

Comment (neutrinoceros) [reply yt/data_objects/data_containers.py]: Almost there. I'll add further comments.

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:714]: we don't want this deprecated parameter to be valid if passed as _positional_, so let's mark it as keyword only
```suggestion
        *,
        JSONdir=None,
```

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:806]: an error should be raised (`TypeError`) if `datadir` is still `None` at this point.

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:803]: no need to specify a removal version, we'll try to keep this for a couple minor versions at least, and possibly longer if it's not a burden to maintain
```suggestion
```

**Review by neutrinoceros — COMMENTED** [review]
follow up comments on the deprecated API part

File: yt/data_objects/data_containers.py
Comment (mtryan83) [reply yt/data_objects/data_containers.py:806]: `None` is an allowed (and is the default) value for `datadir` though. From the Firefly doc/code, a value of `none` defaults to `$HOME/<file_prefix>` (and `file_prefix` defaults to `Data`).

File: yt/data_objects/data_containers.py
Comment (mtryan83) [inline yt/data_objects/data_containers.py:727]:
```suggestion
            path will assume relative to `${HOME}`. A value of `None` will default to `${HOME}/Data`.
```

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [reply yt/data_objects/data_containers.py:806]: Yes that would make sense indeed ! While we're at it, we should probably also document where additional `**kwargs` are being passed to, since that's also supported.

File: yt/data_objects/data_containers.py
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:758]:
```suggestion
        Any additional keyword arguments are passed to
        firefly.data_reader.Reader.__init__
```

**Review by neutrinoceros — DISMISSED** [review]

**Comment by agurvich** [comment]
So, I am only just now tuning in. I am traveling until next Monday 10/3 and won't have a chance to formally review this until then. @mtryan83, first of all, thank you for putting all of this together! from your original PR summary this seems like exactly the sort of changes that I've had on a to-do list to do since we updated the API and submitted the Firefly paper so I am in your debt for handling it. 

It's been hard to follow the updates to the PR since then, has it all mostly been yt documentation stuff? Is there something in particular I should pay close attention to (it sounds like you covered all the bookkeeping stuff like switching JSONDIR and `.ffly` format, etc.). I can have a look at the file diffs but it may be worth coordinating and changing things upstream if there's something worth cleaning up there (e.g. if it means removing a warning on yt's side because I just do better type checking on firefly's side).

**Comment by neutrinoceros** [comment]
>So, I am only just now tuning in. I am traveling until next Monday 10/3 and won't have a chance to formally review this until then.

No rush on our side !

> it may be worth coordinating and changing things upstream if there's something worth cleaning up there (e.g. if it means removing a warning on yt's side because I just do better type checking on firefly's side).

AFAICT there's nothing we'd want to clean up upstream.

**Comment by mtryan83** [comment]
I think the main thing is just making sure I updated the API correctly.  From what I could tell, it was just `JSONdir` --> `datadir`, `tracked_*` --> `field_*`, `Firefly` --> `firefly`, and `dumpToJSON` --> `writeToDisk` and `options` --> `settings` (though those two only appear in the documentation).

I don't see anything to clean up upstream either.

Looks like I missed a JSONdir though, @neutrinoceros, sorry!

**Comment by neutrinoceros** [comment]
For later reference, I previously triaged this for the 4.1 milestone, but I'm going to take it back, because that release is already far behind schedule. I would still be happy to include the patch in the release if we can get a second review in time, which seems likely, I just want to signal that the patch shouldn't be considered blocking.

**Comment by mtryan83** [comment]
(fast-cache and firefly dependencies were _touching_! Couldn't automerge anymore)

**Review by agurvich — APPROVED** [review]
These look great to me, thank you @mtryan83 for cleaning all this up with the new API. I'm hoping that I won't be compelled to change it again any time soon now that we have the .ffly format (which is what inspired most of the changes).

**Review by neutrinoceros — APPROVED** [review]

**Pull request merged**
